src/napi/js_native_ext_api_v8.cpp

```diff
@@ -0,0 +1,273 @@
+
+  *result = new napi_env_scope__(env->isolate, env->context());
+  return napi_ok;
+}
+
+napi_status napi_ext_close_env_scope(napi_env env, napi_env_scope scope) {
+  CHECK_ENV(env);
+  CHECK_ARG(env, scope);
+
+  delete scope;
+  return napi_ok;
+}
+
+napi_status napi_ext_has_unhandled_promise_rejection(
+    napi_env env,
+    bool *result) {
+  CHECK_ENV(env);
+  CHECK_ARG(env, result);
+
+  auto runtime = v8runtime::V8Runtime::GetCurrent(env->context());
+  CHECK_ARG(env, runtime);
+
+  *result = runtime->HasUnhandledPromiseRejection();
+  return napi_ok;
+}
+
+napi_status napi_get_and_clear_last_unhandled_promise_rejection(
+    napi_env env,
+    napi_value *result) {
+  CHECK_ENV(env);
+  CHECK_ARG(env, result);
+
+  auto runtime = v8runtime::V8Runtime::GetCurrent(env->context());
+  CHECK_ARG(env, runtime);
+
+  auto rejectionInfo = runtime->GetAndClearLastUnhandledPromiseRejection();
+  *result =
+      v8impl::JsValueFromV8LocalValue(rejectionInfo->value.Get(env->isolate));
+  return napi_ok;
+}
+
+napi_status napi_ext_run_script(
+    napi_env env,
+    napi_value script,
+    const char *source_url,
+    napi_value *result) {
+  NAPI_PREAMBLE(env);
+  CHECK_ARG(env, script);
+  CHECK_ARG(env, result);
+
+  v8::Local<v8::Value> v8_script = v8impl::V8LocalValueFromJsValue(script);
+
+  if (!v8_script->IsString()) {
+    return napi_set_last_error(env, napi_string_expected);
+  }
+
+  v8::Local<v8::Context> context = env->context();
+
+  v8::Local<v8::String> urlV8String =
+      v8::String::NewFromUtf8(context->GetIsolate(), source_url)
+          .ToLocalChecked();
+  v8::ScriptOrigin origin(urlV8String);
+
+  auto maybe_script = v8::Script::Compile(
+      context, v8::Local<v8::String>::Cast(v8_script), &origin);
+  CHECK_MAYBE_EMPTY(env, maybe_script, napi_generic_failure);
+
+  auto script_result = maybe_script.ToLocalChecked()->Run(context);
+  CHECK_MAYBE_EMPTY(env, script_result, napi_generic_failure);
+
+  *result = v8impl::JsValueFromV8LocalValue(script_result.ToLocalChecked());
+  return GET_RETURN_STATUS(env);
+}
+
+napi_status napi_ext_collect_garbage(napi_env env) {
+  env->isolate->RequestGarbageCollectionForTesting(
+      v8::Isolate::kFullGarbageCollection);
+  return napi_status::napi_ok;
+}
+
+namespace node {
+
+namespace per_process {
+// From node.cc
+// Tells whether the per-process V8::Initialize() is called and
+// if it is safe to call v8::Isolate::GetCurrent().
+bool v8_initialized = false;
+} // namespace per_process
+
+// From node_errors.cc
+[[noreturn]] void Assert(const AssertionInfo &info) {
+  std::string name = "aaa"; // [vmoroz] GetHumanReadableProcessName();
+
+  fprintf(
+      stderr,
+      "%s: %s:%s%s Assertion `%s' failed.\n",
+      name.c_str(),
+      info.file_line,
+      info.function,
+      *info.function ? ":" : "",
+      info.message);
+  fflush(stderr);
+
+  // [vmoroz] Abort();
+}
+
+} // namespace node
+
+// TODO: [vmoroz] verify that finalize_cb runs in JS thread
```

Comment:
>finalize_cb runs in JS [](start = 30, length = 22)

As this is triggered just before finalizing the object, it is highly likely to be called synchrously from the thread where the GC is taking place, which could possibly be one of the background threads.

Reply:
I need to write a unit test here and if the thread assumption is broken, then I need to have a different implementation.

Reply:
I will resolve this TODO in a future PR